### ISSUE TITLE
[RFC] Using main cmssw repository for L1Tracking development

### ISSUE THREAD

**Issue by iarspider** [cms-sw/cmssw#46748]
As discussed during yesterday's [Core Software meeting](https://indico.cern.ch/event/1480583/), we propose the following solution for better integration of L1Tracking development in CMSSW:

* Create a new category (similar to, e.g. `heterogeneous`) and add these pacakges to it: `L1Trigger/TrackFindingTracklet`, `L1Trigger/TrackFindingTMTT`, `L1Trigger/TrackerTFP`,  `L1Trigger/TrackerDTC`, `L1Trigger/TrackTrigger`, `DataFormats/L1TrackTrigger`
    * These packages currently are in `l1` category, we can move them out
* Add corresponding experts to that category
* Add special tests (similar to, e.g., `gpu`) that can be requested in addition to regular tests

This way, when a PR is opened that changes files under these packages, only the relevant experts will be notified, and their signature will be required to merge the PR.

@tomalin FYI

**Comment by smuzaffar** [comment]
type doc

**Comment by smuzaffar** [comment]
> Add special tests (similar to, e.g., gpu) that can be requested in addition to regular tests

@tomalin , for additional `l1track` tests, we need to teach bot how to recognize it and what to run. You can see https://github.com/cms-sw/cms-bot/pull/2354 and https://github.com/cms-sw/cms-bot/pull/2240 as examples for running enable and run extra PR test. @cms-sw/externals-l2 can take care of it if you point us your current test ( run via gitlab ci) Or feel free to open cms-bot PR

**Comment by makortel** [comment]
assign l1

**Comment by tomalin** [comment]
Thanks @smuzaffar . How do I add experts to the new l1track git-group, that will sign off on PRs? 

Our current git CI test works like this:

1) In the L1track group's CMSSW github fork, our github CI https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_2_0_pre4/.github/workflows/github_CI.yml is run when someone makes a PR to our development branch.

2) Our github CI calls the utility https://github.com/cms-L1TK/gitlab-mirror-and-ci-action ,  to trigger a git CI job to run on gitlab.

3) The gitlab CI job compiles & runs our CMSSW L1 track code. It is https://gitlab.cern.ch/cms-l1tk/cmssw_CI/-/blob/masterCI/.gitlab-ci.yml .

It's not obvious to me how to translate (3) for your "bot". I'll be at CERN for CMS week, if talking in person would help?

**Comment by smuzaffar** [comment]
> How do I add experts to the new l1track git-group, that will sign off on PRs?

You can open a PR with update `CMSSW_L2` [here](https://github.com/cms-sw/cms-bot/blob/master/categories.py#L60-L158) and add members responsible for `l1track` or let us know the github usernames and we can update it. Once that change is in then one can explicitly do `assign l1track` to notify `l1track` members about the PR . In this case bot will add `l1track-pending` label which `l1track` members can sign off.

We will look in to your CI script and add those to cms-bot

**Comment by tomalin** [comment]
@smuzaffar Good news! we've now found a volunteer to create this unit test, so work should start soon.

Also, can you please add "tomalin" and "aehart" to the "l1track" category, so they will be asked to sign off on PRs on behalf of the L1 track community?

**Comment by smuzaffar** [comment]
thanks @tomalin .
@iarspider , can you please update cms-bot to add the new category and mapping for packages. Can you please also see how to add tests mentioned in https://github.com/cms-sw/cmssw/issues/46748#issuecomment-2518062901  in to bot

**Comment by tomalin** [comment]
We've just found a PhD student volunteer to do this task of migrating our L1 track git CI tests to a CMSSW unit test.  He'll doubtless be appreciative of any advice that @smuzaffar or @iarspider can give, once he's a little further into the work.

**Comment by smuzaffar** [comment]
sure , just ping us whenever he is ready

**Comment by wlivesay** [comment]
Dear @smuzaffar & @iarspider , I am the student helping work on this L1TK Unit Test Project. I’d be grateful for your advice. I drafted this L1TK unit test script  https://github.com/wlivesay/L1TK_Unit_Test/blob/L1TK-dev-20_1_0_pre1/L1Trigger/TrackFindingTracklet/test/L1TKUnitTest.sh .

1. I registered this unit test in https://github.com/wlivesay/L1TK_Unit_Test/blob/L1TK-dev-20_1_0_pre1/L1Trigger/TrackFindingTracklet/test/BuildFile.xml . Is this enough to make it known to cmsbot, so it can be used during PR review?
2. Should L1TKUnitTest.sh be modified to use https://github.com/cms-sw/cms-bot/blob/master/pr_testing/run-pr-unittests.sh ? If so, how and for what?
3. We ideally want to run 3 algorithms over 1000 MC events at 200 pileup, in order to check tracking performance hasn’t changed. Is this possible? (We currently run this on CERN gitlab CI).
4. How should we access the input RelVal MC datasets from CMSSW? (Our git CI accesses skimmed MC from CERNbox). Surely we shouldn’t generate the datasets on the fly (e.g. With runTheMatrix workflow = 34634.78 TTbar_14TeV+Run4D121PU_L1Complete)?

Thanks for your help with this

**Comment by tomalin** [comment]
Dear @iarspider and @smuzaffar we really need your advice on Billy's query above, so we can finish our L1 track unit test. What do you think?

**Comment by smuzaffar** [comment]
ah sorry @wlivesay , I missed your comment here.
About your point 1,2: The links in your comment are broken ( at least I can not access them) but if unit test is already mentioned in `cmssw/<sub-system>/<package>/test/BuildFile.xml` then you do not need any thing else. For PR tests, SCRAM will run it when `<sub-system>/<package>` is checkout and for IBs scram will always run it. 

About 3: How long it will take to run these  "3 algorithms over 1000 MC events at 200 pileup". Full unit tests currently run under 45mins. We do not want to have long running (taking hours) tests for PRs. I suggest that for normal PR/IBs tests we run small number of tests and additionally we can dd a seperate unit test to run  "3 algorithms over 1000 MC events at 200 pileup" on demand ( e.g one can comment on PR `enable l1tk` and then bot can run an additional test .

About 4: If input dataset is accessible via xrootd then our system will automatically cache the accessed root files un a dedicated eos cache directory. By the way, how big is this dataset?

**Comment by iarspider** [comment]
@wlivesay @tomalin  just FYI - I am no longer working at CMS, so can't help you with this project in any way. Sorry!

**Comment by smuzaffar** [comment]
@wlivesay , ok the two root files this tests uses are 38GB each. How often do we change these files? Are they known to das client ? I mean were these files produced via some official production and registered in DAS ? If yes then you can just use the logical file name of the files , first time these will be download  with xrootd and then our bot will cache then in CERN EOS.

If these are not known to DAS then we can copy then manual under EOS and then you can use their LFN in test

**Comment by wlivesay** [comment]
@smuzaffar, thank you for your responses. For the full 1000 events running each algorithm in parallel it takes less than 1 hour. Do you think we can reduce the number of events to fit within your expected unit test duration? Perhaps we can try 100 events? Am I correct in thinking if we decide to run this automatically I would keep the line in BuildFile.xml and there would be no need for a command (enable l1tk) from our end each time?

Xrootd is available and the datasets we use can be found on DAS, so that shouldn't be a problem. The datasets on DAS are a bit larger (the file I sent you uses slightly lighter skims of the datasets) so maybe 125GB for 1000 events. We would likely change these datasets about once a year (that's what we currently do). Whose responsibility is this generally to update the datasets?

**Comment by smuzaffar** [comment]
@wlivesay , Yes I think it is better to run for 100 events (or even less) as a unit tests. Unit tests automatically run for IBs (always) and PR tests ( if package where unit tests is defined has changes or packages which it depends on have interface changes). Also as `SCRAM` runs unit tests in parallel (N tests on a N cores machine) so ideally we want each test to use 1 process otherwise system will be overloaded. So running even 100 evts in serial might take over an hour, so I suggest the unit test should run few event in serial.

I was also suggesting that for PR tests we additionally have a separate Jenkins job which runs your configuration in parallel for 1000 events. This will be an additional tests which you can request by adding e.g. `enable l1tk` comment on the PR for which you want to run the full test. We can also setup bot to run this full test if PR contains direct changes in packages belonging to `l1track`

**Comment by wlivesay** [comment]
@smuzaffar, I think I understand now. I thought 100 events would take only ~15 mins like it does for me locally (in series), but it seems to be different for your end. In this case, I like your suggestion of automatically running few events (~5?) to test everything (compilation, dataset access, code style checks, etc) except for the performance and then running the command for our 1000 event performance check when necessary.

If we do this, I would keep a light unit test in BuildFile.xml, but how would "enable l1tk" know where my heavy unit test is? How would I add a unit test to be recognized by that command?

**Comment by smuzaffar** [comment]
> If we do this, I would keep a light unit test in BuildFile.xml, but how would "enable l1tk" know where my heavy unit test is? How would I add a unit test to be recognized by that command?

@wlivesay , if you make unit test (some bash script) configurable e.g. `l1tk-test.sh -e <num-events> -j <parallel-jobs>` then in the BuilsFile.xml you can use `l1tk-test.sh -e 10 -j 1` and for additional test a dedicated jenkins job can use the same script but with high number of events `l1tk-test.sh -e 1000 -j $(nproc)`.  We will take care of adding jenkins job to run the additional test and teach cms-bot to recognize `enable l1tk` comment

**Comment by wlivesay** [comment]
@smuzaffar Sure, I can work on this. One issue I have is my threshold values are hard-coded currently and would require manual updating. Is there a simple way to update my threshold values with the values the current state of the code produce? I would like to always compare the new PR to the current values to flag any jumps if possible. Do you know if this is this done by any other groups? Is there some sort of command we could run that would "update L1TK unit test" and change stored "newPR thresholds" to become the "current code thresholds" inside my unit test? If this is not a simple ask, I am sure we would be comfortable with keeping a general threshold and update it every 6-12 months (as we do now).

**Comment by wlivesay** [comment]
I made a new commit to my repo with changes. It now accepts the 2 inputs you suggested and I checked that it works locally for both (10 events 1 job) and (1000 events 1 job). I have it skip the threshold validation for any event count less than 1000 as it will surely fail due to poor statistics with my hard-coded values corresponding to a 1000 event run.

I currently have the code store temporary outputs (required for our code to run in full) in ...test/unit_test_results/HYBRID_run for example. Is this fine for when this is run on jenkins? I was assuming the files could help debug if the unit test fails somewhere along the way.

**Comment by smuzaffar** [comment]
@wlivesay, note that `CMSSW/src` is not writable during unit tests. During automatic testing we do not allow tests to write any thing under `CMSSW/src` . So if you are explicitly creating output under your `CMSSw/src/SubSystem/Package/test/unit_test_results` path tne unit test will fail.  

When we run `scram build runtests` then each test is started from `$CMSSW_BASE/unit_tests/<test-name>` . So as long as you create `unit_test_results` in your `cwd` then things will work. For IBs and CI tests we only make the `stdout/stderr` of the test available. Any other output generated during the test is discarding. But for special additional `l1tk` test we can upload the generated output to `cmssdt` server. Any idea how much data/output files the full 1000 generate? 

For comparing the new PR results, I assume you want a reference output with default values/thresholds .. right? And the additional full 1000 events job needs access to this ref for comparison... right? Yes we have some CI tests which does this e.g. our normal DQM comparison and HLT P2 timing jobs does this. 

Are you at CERN, if yes then we can meet sometime next week to go though all these. Otherwise we can have a zoom chat to sort out all these details
